### PR TITLE
Implement "in memory" backend

### LINKED ISSUE #6: Add "In Memory" cache backend.
This looks like a solid cache tool, but lack of an in memory backend to both avoid filesystem interaction and not require an external dependency like running mongodb, is a pretty big hinderance for me.

### PR DESCRIPTION
**Pull request by cthoyt**
Closes #6 

This PR implements an "in memory" backend. It mostly excludes all of the locking and waiting functionalities that were in the pickle backend, since the in-memory should have negligible write and read times (in comparison to pickle). I may have misunderstood the purpose of that functionality, so if it's needed, I think it will be easy to put back in. 

This PR also includes the changes from #38 because those were necessary to enable switching between multiple backends.

**Caveat** the tests aren't written in a very reusable way, so I'm not sure the best way to test this without duplicating a lot of code. I'd appreciate some advice/help on what to do here

### CODE REVIEW

File: cachier/memory_core.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+    def mark_entry_not_calculated(self, key):
+        try:
+            self.cache[key]['being_calculated'] = False
+        except KeyError:
+            pass  # that's ok, we don't need an entry in that case
+
+    def wait_on_entry_calc(self, key):
+        entry = self.cache[key]
+        # I don't think waiting is necessary for this one
```
Comment (cthoyt) [inline cachier/memory_core.py:58]: it turns out this assumption isn't at all correct 😆

File: cachier/core.py [hunk 2]
```diff
@@ -110,6 +116,10 @@ def cachier(
         A callable that takes no arguments and returns a pymongo.Collection
         object with writing permissions. If unset a local pickle cache is used
         instead.
+    backend : str, optional
+        The name of the backend to use. If None, defaults to 'mongo' when
+        the ``mongetter`` argument is passed, otherwise defaults to 'pickle'.
+        Valid options currently include 'pickle' and 'mongo'.
```
Comment (shaypal5) [inline cachier/core.py:122]: Please add 'memory' as a valid option.

**Comment by shaypal5** [comment]
Yes, you need locking for thread safety. :)

I'm pulling this into a branch here and I'll add the locking functionality myself + README additions.

**Comment by cthoyt** [comment]
@shaypal5 the change of merge target was a bit unexpected, I think it's much better to just push to my branch for consistency. Anyway I hope that my commits don't get squashed and lost in the history once you finish it :)

**Comment by shaypal5** [comment]
I just didn't want to have to wait for you to approve PRs from me to your branch, to then just have these as part of this PR.

That sounds like a vicious cycle of waiting. I just wanted to get this done. :)

**Comment by shaypal5** [comment]
And I AM going to squash all your commits into a single one, just for a more readable history.

I've started to use dynamic rebasing a lot recently, and can finally appreciate how important it can be to maintain clean commit history. :)

**Comment by cthoyt** [comment]
Nice! I think GitHub allows the repo owner to make commits to branches that are in PRs back to the upstream. I've done this before too

**Comment by shaypal5** [comment]
Released in `v1.5.0` !!! 🥇 🥳

Congrats and great work, @cthoyt ! First major outside contribution to this project!
I'm so proud. May it be the first of many (by you and/or users of the package)!

**Pull request merged**